File: ios/Classes/DocumentScanner/FlutterDocumentScanner.swift

import AVFoundation
import EkycID
import Foundation

public class FlutterDocumentScanner: NSObject, FlutterPlatformView, DocumentScannerEventListener {
    let frame: CGRect
    let viewId: Int64
    var flutterScannerView: UIView?
    var scanner: DocumentScannerView!
    var methodChannel: FlutterMethodChannel?
    var eventChannel: FlutterEventChannel?
    var eventStreamHandler: DocumentScannerEventStreamHandler?

    init(frame: CGRect, viewId: Int64, messenger: FlutterBinaryMessenger, args: Any?) {
        let screenSize = UIScreen.main.bounds
        let screenWidth = screenSize.width
        let screenHeight = screenSize.height
        
        print(frame.width)
        print(frame.height)
        
        self.frame = frame

        self.viewId = viewId

        super.init()
        self.flutterScannerView = UIView(frame: CGRect(x: 0, y: 0, width: screenWidth, height: screenHeight))
        self.methodChannel = FlutterMethodChannel(name: "DocumentScanner_MethodChannel_" + String(viewId), binaryMessenger: messenger)
        self.eventChannel = FlutterEventChannel(name: "DocumentScanner_EventChannel_" + String(viewId), binaryMessenger: messenger)
        self.eventStreamHandler = DocumentScannerEventStreamHandler()
        self.methodChannel!.setMethodCallHandler(self.onMethodCall)
        self.eventChannel!.setStreamHandler(self.eventStreamHandler)
    }

    public func view() -> UIView {
        return self.flutterScannerView!
    }

    private func start(call: FlutterMethodCall, result: @escaping FlutterResult) throws {
        let args = call.arguments as! [String: Any?]
        self.scanner = DocumentScannerView(frame: self.flutterScannerView!.frame)
        self.scanner.addListener(self)
        self.scanner.autoresizingMask = [.flexibleWidth, .flexibleHeight]
        self.flutterScannerView!.addSubview(self.scanner)
        self.scanner.start(options: self.argsToDocumentScannerOptions(args))
        result(true)
    }
    
    private func argsToDocumentScannerOptions(_ args: [String: Any?]) -> DocumentScannerOptions {
        let cameraOptions = args["cameraOptions"] as! [String: Any?]
        let captureDurationCountDown = cameraOptions["captureDurationCountDown"] as! Int
        let faceCropScale = cameraOptions["faceCropScale"] as! NSNumber
        let roiSize = cameraOptions["roiSize"] as! NSNumber
        let minDocWidthPercentage = cameraOptions["minDocWidthPercentage"] as! NSNumber
        let maxDocWidthPercentage = cameraOptions["maxDocWidthPercentage"] as! NSNumber
        let scannableDocuments = args["scannableDocuments"] as! [[String: Any]]
        return DocumentScannerOptions(
            cameraOptions: DocumentScannerCameraOptions(
                captureDurationCountDown: captureDurationCountDown,
                faceCropScale: faceCropScale.floatValue,
                roiSize: roiSize.floatValue,
                minDocWidthPercentage: minDocWidthPercentage.floatValue,
                maxDocWidthPercentage: maxDocWidthPercentage.floatValue
            ),
            scannableDocuments: scannableDocuments.map { e in
                var doc = e as! [String: String?]
                var mainSide = doc["mainSide"]!
                var secondarySide = doc["secondarySide"]!
                
                return ScannableDocument(
                    mainSide: StringToObjectDetectionObjectTypeMapping[mainSide!]!,
                    secondarySide: secondarySide != nil ? StringToObjectDetectionObjectTypeMapping[secondarySide!]! : nil
                )
            }
        )
    }
    
    public func onInitialized() {
        self.eventStreamHandler?.sendOnInitializedEventToFlutter()
    }
    
    public func onDocumentScanned(mainSide: EkycID.DocumentScannerResult, secondarySide: EkycID.DocumentScannerResult?) {
        self.eventStreamHandler?.sendOnDocumentScannedEventToFlutter(mainSide: mainSide, secondarySide: secondarySide)
    }
    
    public func onFrameStatusChanged(_ frameStatus: EkycID.FrameStatus) {
        self.eventStreamHandler?.sendOnFrameStatusChangedEventToFlutter(frameStatus)
    }
    
    public func onCurrentSideChanged(_ currentSide: EkycID.DocumentSide) {
        self.eventStreamHandler?.sendOnCurrentSideChangedEventToFlutter(currentSide)
    }
    
    public func onCaptureCountDownChanged(current: Int, max: Int) {
        self.eventStreamHandler?.sendOnCaptureCountDownChangedEventToFlutter(current: current, max: max)
    }
    
    private func nextImage(call: FlutterMethodCall, result: @escaping FlutterResult) throws {
        if self.scanner != nil {
            self.scanner.nextImage()
        }
        result(true)
    }

    private func dispose(call: FlutterMethodCall, result: @escaping FlutterResult) throws {
        if (self.scanner != nil) {
            self.scanner!.stop()
            self.scanner = nil
        }
        result(true)
    }
    
    private func reset(call: FlutterMethodCall, result: @escaping FlutterResult) throws {
        if (self.scanner != nil) {
            self.scanner!.reset()
        }
        result(true)
    }

    private func onMethodCall(call: FlutterMethodCall, result: @escaping FlutterResult) {
        switch call.method {
        case "start":
            do {
                try self.start(call: call, result: result)
            } catch {
                result(FlutterError(code: "initialize Error", message: nil, details: nil))
            }
        case "dispose":
            do {
                try self.dispose(call: call, result: result)
            } catch {
                result(FlutterError(code: "dispose Error", message: nil, details: nil))
            }
            break
        case "nextImage":
            do {
                try self.nextImage(call: call, result: result)
            } catch {
                result(FlutterError(code: "dispose Error", message: nil, details: nil))
            }
            break
        case "reset":
            do {
                try self.reset(call: call, result: result)
            } catch {
                result(FlutterError(code: "dispose Error", message: nil, details: nil))
            }
            break
        default:
            result(FlutterMethodNotImplemented)
        }
    }
    
    class DocumentScannerEventStreamHandler: NSObject, FlutterStreamHandler {
        var events: FlutterEventSink?

        func onListen(withArguments arguments: Any?, eventSink events: @escaping FlutterEventSink) -> FlutterError? {
            print("onListen")
            self.events = events
            return nil
        }

        func onCancel(withArguments arguments: Any?) -> FlutterError? {
            print("onCancel")
            return nil
        }

        func sendOnDocumentScannedEventToFlutter(_ mainside: DocumentScannerResult, _ secondarySide: DocumentScannerResult?) {
            print("sendOnDocumentScannedEventToFlutter")
            if self.events != nil {
                DispatchQueue.main.async {
                    var event = [String: Any]()
                    event["type"] = "onInitialized"
                    self.events!(event)
                }
            }
        }
        
        func sendOnFrameStatusChangedEventToFlutter(_ frameStatus: FrameStatus) {
            if (self.events != nil) {
                DispatchQueue.main.async {
                    var event = [String: Any]()
                    event["type"] = "onFrameStatusChanged"
                    event["values"] = "\(frameStatus)"
                    self.events!(event)
                }
            }
        }
        
        func sendOnCurrentSideChangedEventToFlutter(_ currentSide: EkycID.DocumentSide) {
            if (self.events != nil) {
                DispatchQueue.main.async {
                    var event = [String: Any]()
                    event["type"] = "onCurrentSideChanged"
                    event["values"] = "\(currentSide)"
                    self.events!(event)
                }
            }
        }
        
        func sendOnDocumentScannedEventToFlutter(mainSide: EkycID.DocumentScannerResult, secondarySide: EkycID.DocumentScannerResult?) {
            if (self.events != nil) {
                DispatchQueue.main.async {
                    var event = [String: Any]()
                    event["type"] = "onDocumentScanned"
                    
                    var result = [String: Any?]()
                    result["mainSide"] = mainSide.toFlutterMap()
                    if secondarySide != nil {
                        result["secondarySide"] = secondarySide!.toFlutterMap()
                    }
                    
                    event["values"] = result
                    self.events!(event)
                }
            }
        }
        
        func sendOnCaptureCountDownChangedEventToFlutter(current: Int, max: Int) {
            if (self.events != nil) {
                DispatchQueue.main.async {
                    var event = [String: Any]()
                    event["type"] = "onCaptureCountDownChanged"
                    var result = [String: Any?]()
                    result["current"] = current
                    result["max"] = max
                    event["values"] = result
                    self.events!(event)
                }
            }
        }
    }
}

extension DocumentScannerResult {
    func toFlutterMap() -> [String: Any?] {
        var values = [String: Any?]()
        values["documentType"] = "\(self.documentType)"
        values["documentGroup"] = "\(self.documentGroup)"
        values["fullImage"] = FlutterStandardTypedData(bytes: self.fullImage.jpegData(compressionQuality: 0.8)!)
        values["documentImage"] = FlutterStandardTypedData(bytes: self.documentImage.jpegData(compressionQuality: 0.8)!)
        if self.faceImage != nil {
            values["faceImage"] = FlutterStandardTypedData(bytes: self.faceImage!.jpegData(compressionQuality: 0.8)!)
        } else {
            values["faceImage"] = nil
        }

        return values
    }
}

File: android/src/main/kotlin/com/ekycsolutions/ekyc_id_flutter/DocumentScanner/FlutterDocumentScanner.kt
package com.ekycsolutions.ekyc_id_flutter.DocumentScanner

import android.annotation.SuppressLint
import android.app.Activity
import android.content.Context
import android.graphics.Bitmap
import android.os.Build
import android.util.Log
import android.view.LayoutInflater
import android.view.View
import android.widget.LinearLayout
import androidx.annotation.RequiresApi
import com.ekycsolutions.ekyc_id_flutter.R
import com.ekycsolutions.ekycid.core.models.FrameStatus
import com.ekycsolutions.ekycid.core.objectdetection.ObjectDetectionObjectTypeStringToObjectTypeMapping
import com.ekycsolutions.ekycid.documentscanner.DocumentScannerEventListener
import com.ekycsolutions.ekycid.documentscanner.DocumentScannerOptions
import com.ekycsolutions.ekycid.documentscanner.DocumentScannerResult
import com.ekycsolutions.ekycid.documentscanner.DocumentScannerView
import com.ekycsolutions.ekycid.documentscanner.DocumentSide
import com.ekycsolutions.ekycid.documentscanner.ScannableDocument
import com.ekycsolutions.ekycid.documentscanner.cameraview.DocumentScannerCameraOptions
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.plugin.common.EventChannel
import io.flutter.plugin.common.MethodCall
import io.flutter.plugin.common.MethodChannel
import io.flutter.plugin.platform.PlatformView
import java.io.ByteArrayOutputStream
import java.util.*
import kotlin.collections.ArrayList
import kotlin.collections.HashMap


class FlutterDocumentScanner(
    private var binding: FlutterPlugin.FlutterPluginBinding,
    private var context: Context,
    private val viewId: Int
): PlatformView, MethodChannel.MethodCallHandler, DocumentScannerEventListener {
    private var scanner: DocumentScannerView? = null

    private var scannerView: LinearLayout = LinearLayout(context)
    private val methodChannel: MethodChannel =
        MethodChannel(binding.binaryMessenger, "DocumentScanner_MethodChannel_$viewId")
    private val eventChannel: EventChannel =
        EventChannel(binding.binaryMessenger, "DocumentScanner_EventChannel_$viewId")
    private val eventStreamHandler = DocumentScannerEventStreamHandler(context)

    init {
        this.methodChannel.setMethodCallHandler(this)
        this.eventChannel.setStreamHandler(eventStreamHandler)
    }


    override fun getView(): View {
        return scannerView
    }

    override fun dispose() {

    }

    override fun onMethodCall(call: MethodCall, result: MethodChannel.Result) {
        when (call.method) {
            "start" -> {
                start(call, result)
            }
            "nextImage" -> {
                nextImage(call, result)
            }
            "dispose" -> {
                disposeFlutter(call, result)
            }
            "reset" -> {
                reset(call, result)
            }
            else -> {
                result.notImplemented()
            }
        }
    }

    private fun start(call: MethodCall, result: MethodChannel.Result) {
        Log.d("START","Processing is starting")
        try {
            val args = call.arguments as HashMap<*, *>
            val cameraOptions = args["cameraOptions"] as HashMap<String, Any>
            val scannableDocuments = args["scannableDocuments"] as ArrayList<HashMap<String, Any>>

            this.scanner = DocumentScannerView(context)
            this.scanner!!.addListener(this)
            this.scanner!!.start(
                DocumentScannerOptions(
                    DocumentScannerCameraOptions(
                        cameraOptions["captureDurationCountDown"] as Int,
                        (cameraOptions["faceCropScale"] as Double).toFloat(),
                        (cameraOptions["roiSize"] as Double).toFloat(),
                        (cameraOptions["minDocWidthPercentage"] as Double).toFloat(),
                        (cameraOptions["maxDocWidthPercentage"] as Double).toFloat(),
                    ),
                    ArrayList(scannableDocuments.map {
                        ScannableDocument(
                            ObjectDetectionObjectTypeStringToObjectTypeMapping[it["mainSide"]!!]!!,
                            if (it["secondarySide"] != null) ObjectDetectionObjectTypeStringToObjectTypeMapping[it["secondarySide"]!!] else null,
                        )
                    }),
                )
            )

            this.scanner!!.layoutParams = LinearLayout.LayoutParams(
                LinearLayout.LayoutParams.MATCH_PARENT,
                LinearLayout.LayoutParams.MATCH_PARENT
            )
            this.scannerView.addView(this.scanner)
            result.success(true)
        } catch (e: Exception) {
            result.error(e.toString(), e.message, "")
        }
    }

    private fun disposeFlutter(call: MethodCall, result: MethodChannel.Result) {
        try {
            if (this.scanner != null) {
                this.scanner!!.stop()
                this.scanner = null
            }
            result.success(true)
        } catch (e: Exception) {
            result.error(e.toString(), e.message, "")
        }
    }

    private fun nextImage(call: MethodCall, result: MethodChannel.Result) {
        try {
            if (this.scanner != null) {
                this.scanner!!.nextImage()
            }
            result.success(true)
        } catch (e: Exception) {
            result.error(e.toString(), e.message, "")
        }
    }

    private fun reset(call: MethodCall, result: MethodChannel.Result) {
        try {
            if (this.scanner != null) {
                this.scanner!!.reset()
            }
            result.success(true)
        } catch (e: Exception) {
            result.error(e.toString(), e.message, "")
        }
    }

    override fun onInitialized() {
        this.eventStreamHandler?.sendOnInitializedEventToFlutter()
    }

    @RequiresApi(Build.VERSION_CODES.N)
    override fun onDocumentScanned(mainSide: DocumentScannerResult, secondarySide: DocumentScannerResult?) {
        this.eventStreamHandler?.sendOnDocumentScannedEventToFlutter(mainSide, secondarySide)
    }

    override fun onFrameStatusChanged(frameStatus: FrameStatus) {
        this.eventStreamHandler?.sendOnFrameStatusChangedEventToFlutter(frameStatus)
    }

    override fun onCaptureCountDownChanged(current: Int, max: Int) {
        this.eventStreamHandler?.sendOnCaptureCountDownChangedEventToFlutter(current, max)
    }

    override fun onCurrentSideChanged(currentSide: DocumentSide) {
        this.eventStreamHandler?.sendOnCurrentSideChangedEventToFlutter(currentSide)
    }

    class DocumentScannerEventStreamHandler(private var context: Context) : EventChannel.StreamHandler {
        private var events: EventChannel.EventSink? = null

        override fun onListen(arguments: Any?, events: EventChannel.EventSink?) {
            this.events = events
        }

        override fun onCancel(arguments: Any?) {
            this.events = null
        }

        fun sendOnInitializedEventToFlutter() {
            if (events != null) {
                (context as Activity).runOnUiThread {
                    val event = HashMap<String, Any>()
                    event["type"] = "onInitialized"
                    val values = HashMap<String, Any?>()
                    event["values"] = values
                    events!!.success(event)
                }
            }
        }

        fun sendOnFrameStatusChangedEventToFlutter(frameStatus: FrameStatus) {
            if (events != null) {
                (context as Activity).runOnUiThread {
                    val event = HashMap<String, Any>()
                    event["type"] = "onFrameStatusChanged"
                    event["values"] = frameStatus.name
                    events!!.success(event)
                }
            }
        }

        @RequiresApi(Build.VERSION_CODES.N)
        fun sendOnDocumentScannedEventToFlutter(mainSide: DocumentScannerResult, secondarySide: DocumentScannerResult?) {
            if (events != null) {
                (context as Activity).runOnUiThread {
                    val event = HashMap<String, Any>()
                    event["type"] = "onDocumentScanned"
                    val values = HashMap<String, Any?>()
                    values["mainSide"] = documentScannerResultToFlutterMap(mainSide)
                    if (secondarySide!=null) {
                        values["secondarySide"] = documentScannerResultToFlutterMap(secondarySide!!)
                    }

                    event["values"] = values
                    events!!.success(event)
                }
            }
        }


        fun sendOnCurrentSideChangedEventToFlutter(currentSide: DocumentSide) {
            if (events != null) {
                (context as Activity).runOnUiThread {
                    val event = HashMap<String, Any>()
                    event["type"] = "onCurrentSideChanged"
                    event["values"] = currentSide.name
                    events!!.success(event)
                }
            }
        }

        fun sendOnCaptureCountDownChangedEventToFlutter(current: Int, max: Int) {
            if (events != null) {
                (context as Activity).runOnUiThread {
                    val event = HashMap<String, Any>()
                    event["type"] = "onCaptureCountDownChanged"
                    val values = HashMap<String, Any?>()
                    values["current"] = current
                    values["max"] = max
                    event["values"] = values
                    events!!.success(event)
                }
            }
        }

        @RequiresApi(Build.VERSION_CODES.N)
        private fun documentScannerResultToFlutterMap(detection: DocumentScannerResult): HashMap<String, Any?> {
            val values = HashMap<String, Any?>()
            val temp = detection.toMap(context, saveImage = false)
            values["documentType"] = temp["documentType"]
            values["documentGroup"] = temp["documentGroup"]
            values["fullImage"] = bitmapToFlutterByteArray(temp["fullImage"] as Bitmap)
            values["documentImage"] = bitmapToFlutterByteArray(temp["documentImage"] as Bitmap)
            if (detection.faceImage != null) {
                values["faceImage"] = bitmapToFlutterByteArray(temp["faceImage"] as Bitmap)
            } else {
                values["faceImage"] = null
            }
            return values
        }

        private fun bitmapToFlutterByteArray(image: Bitmap): ByteArray {
            val stream = ByteArrayOutputStream()
            image.compress(Bitmap.CompressFormat.JPEG, 90, stream)
            return stream.toByteArray()
        }
    }
}